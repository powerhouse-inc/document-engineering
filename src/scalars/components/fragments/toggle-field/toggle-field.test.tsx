import { fireEvent, screen, waitFor } from '@testing-library/react'
import { describe, it, vi } from 'vitest'
import { renderWithForm } from '../../../lib/testing.js'
import { ToggleField } from './toggle-field.js'

describe('ToggleField Component', () => {
  const mockOnChange = vi.fn()

  it('should match snapshot', () => {
    const { container } = renderWithForm(<ToggleField name="test" label="Test Label" />)
    expect(container).toMatchSnapshot()
  })

  it('should render default status without label', () => {
    renderWithForm(<ToggleField name="test" />)
    expect(screen.queryByText('Test Label')).not.toBeInTheDocument()
  })

  it('should render with label when label prop is provided', () => {
    renderWithForm(<ToggleField name="test" label="Test Label" />)
    expect(screen.getByText('Test Label')).toBeInTheDocument()
  })

  it('should render checked status without label', () => {
    renderWithForm(<ToggleField name="test" value={true} />)
    const toggle = screen.getByRole('switch')
    expect(toggle).toHaveAttribute('data-state', 'checked')
    expect(screen.queryByText('Test Label')).not.toBeInTheDocument()
  })

  it('should render unchecked status with label', () => {
    renderWithForm(<ToggleField name="test" label="Test Label" value={false} />)
    const toggle = screen.getByRole('switch')
    expect(toggle).toHaveAttribute('data-state', 'unchecked')
    expect(screen.getByText('Test Label')).toBeInTheDocument()
  })

  it('should display an error message when errors prop is provided', async () => {
    renderWithForm(<ToggleField name="test" label="Test Label" errors={['Error message']} />)
    await waitFor(() => {
      expect(screen.getByText('Error message')).toBeInTheDocument()
    })
  })

  it('should call onChange when clicked', () => {
    renderWithForm(<ToggleField name="test" label="Test Label" onChange={mockOnChange} />)
    const toggleInput = screen.getByRole('switch')

    fireEvent.click(toggleInput)
    expect(toggleInput).toBeInTheDocument()
    expect(mockOnChange).toHaveBeenCalledTimes(1)
  })

  it('should disable the toggle when disabled prop is true', () => {
    renderWithForm(<ToggleField name="test" disabled />)
    const toggle = screen.getByRole('switch')
    expect(toggle).toBeDisabled()
  })

  it('should render with custom className', () => {
    renderWithForm(<ToggleField name="test" className="custom-class" />)
    const toggle = screen.getByTestId('custom-class')
    expect(toggle).toHaveClass('custom-class')
  })
})
