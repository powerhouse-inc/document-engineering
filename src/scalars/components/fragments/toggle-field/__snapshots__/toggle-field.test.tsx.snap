// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ToggleField Component > should match snapshot 1`] = `
<div>
  <form
    id=":r0:"
    novalidate=""
  >
    <div
      class="flex flex-col gap-1"
      data-testid="custom-class"
    >
      <div
        class="flex items-center"
      >
        <button
          aria-checked="true"
          aria-labelledby=":r1:-label"
          aria-required="false"
          class="focus-visible:ring-ring focus-visible:ring-offset-background peer inline-flex h-4 w-8 shrink-0 cursor-pointer items-center rounded-full border-2 border-transparent transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 data-[state=checked]:bg-[#343839] data-[state=unchecked]:bg-[#D7D8D9] dark:data-[state=checked]:bg-[#3BBE5F] dark:data-[state=unchecked]:bg-[#404446]"
          data-state="checked"
          id=":r1:"
          role="switch"
          type="button"
          value="on"
        >
          <span
            class="peer pointer-events-none block size-3.5 rounded-full bg-[#FCFCFC] shadow-lg ring-0 transition-transform data-[state=checked]:translate-x-3.5 data-[state=unchecked]:translate-x-0"
            data-state="checked"
          />
        </button>
        <input
          aria-hidden="true"
          checked=""
          name="test"
          style="transform: translateX(-100%); position: absolute; pointer-events: none; opacity: 0; margin: 0px; width: 0px; height: 0px;"
          tabindex="-1"
          type="checkbox"
          value="on"
        />
        <label
          class="inline-flex items-center text-sm font-medium leading-[22px] text-gray-900 dark:text-gray-400 group-hover:text-gray-900 dark:group-hover:text-slate-50 ml-2"
          for=":r1:"
          id=":r1:-label"
        >
          Test Label
        </label>
      </div>
    </div>
  </form>
</div>
`;
